coverage/src/lib.rs

```diff
@@ -1,12 +1,112 @@
 // Licensed under the Apache-2.0 license
 
 use anyhow::Context;
+use bit_vec::BitVec;
 use caliptra_builder::{build_firmware_elf, FwId};
 use elf::endian::AnyEndian;
 use elf::ElfBytes;
-use std::collections::HashSet;
+use rnglib::{Language, RNG};
+use std::collections::hash_map::DefaultHasher;
+use std::collections::{HashMap, HashSet};
 use std::fs::File;
-use std::io::{BufRead, BufReader};
+use std::hash::Hasher;
+use std::io::{BufRead, BufReader, BufWriter, Write};
+use std::path::{Path, PathBuf};
+
+pub struct CoverageMap {
+    pub map: HashMap<u64, BitVec>,
+}
+
+impl CoverageMap {
+    pub fn new(paths: Vec<PathBuf>) -> Self {
+        let mut map = HashMap::<u64, BitVec>::default();
+        for path in paths {
+            let new_entry = get_entry_from_path(&path);
+            if map.contains_key(&new_entry.0) {
+                let mut res = map.get(&new_entry.0).unwrap().clone();
+                res.or(&new_entry.1);
+                map.insert(new_entry.0, res);
+            } else {
+                map.insert(new_entry.0, new_entry.1);
+            }
+        }
+        Self { map }
+    }
+}
+pub struct CoverageMapEntry(u64, BitVec);
+pub fn get_entry_from_path(path: &PathBuf) -> CoverageMapEntry {
+    let filename = path.file_name().unwrap().to_str().unwrap();
+    let tag = filename
+        .split('-')
+        .nth(1)
+        .unwrap()
+        .strip_suffix(".bitvec")
+        .unwrap()
+        .parse()
+        .unwrap();
+    let bitmap = read_bitvec_from_file(path).unwrap();
+    CoverageMapEntry(tag, bitmap)
+}
+
+pub fn dump_emu_coverage_to_file(tag: u64, bitmap: &BitVec) -> std::io::Result<()> {
+    let rng = RNG::try_from(&Language::Elven).unwrap();
+
+    let mut filename = rng.generate_name();
+    filename.push_str(&'-'.to_string());
+    filename.push_str(&tag.to_string());
+    filename.push_str(".bitvec");
+
+    let path = std::path::Path::new("/tmp").join(filename);
+
+    let file = File::create(path)?;
+    let mut writer = BufWriter::new(file);
+    serde_json::to_writer(&mut writer, &bitmap)?;
```

Comment:
json is a bit heavyweight for this, no? Doesn't have to be in this PR, we should consider doing something like

```
use zerocopy::AsBytes;

std::fs::write(path, bitmap.as_raw_slice().as_bytes())?;
```

Reply:
Sure. We can do raw binary dumps as a follow up.

Reply:
Reading that back might be a bit tricky; simplest thing would be to use BitVec<u8> instead of BitVec<usize> (the default). Then this becomes:

```
std::fs::write(path, bitmap.as_raw_slice())?;
```

```
let bitmap = BitVec::from_vec(std::fs::read(path)?);
```

Reply:
I will add it to the follow-up PR.